server/cluster_worker.go

```diff
@@ -364,229 +86,166 @@ func (c *raftCluster) HandleAskChangePeer(request *pdpb.AskChangePeerRequest) er
 		log.Infof("region %d peer number equals %d, no need to change peer", regionID, maxPeerNumber)
-		return nil
+		return nil, nil
 	} else if peerNumber < maxPeerNumber {
 		log.Infof("region %d peer number %d < %d, need to add peer", regionID, peerNumber, maxPeerNumber)
 		changeType = raftpb.ConfChangeType_AddNode
 		if peer, err = c.handleAddPeerReq(region); err != nil {
-			return errors.Trace(err)
+			return nil, errors.Trace(err)
 		}
 	} else {
 		log.Infof("region %d peer number %d > %d, need to remove peer", regionID, peerNumber, maxPeerNumber)
 		changeType = raftpb.ConfChangeType_RemoveNode
-		if peer, err = c.handleRemovePeerReq(region, request.Leader); err != nil {
-			return errors.Trace(err)
+		if peer, err = c.handleRemovePeerReq(region, leaderID); err != nil {
+			return nil, errors.Trace(err)
 		}
 	}
 
-	changePeer := &raft_cmdpb.AdminRequest{
-		CmdType: raft_cmdpb.AdminCmdType_ChangePeer.Enum(),
-		ChangePeer: &raft_cmdpb.ChangePeerRequest{
-			ChangeType: changeType.Enum(),
-			Peer:       peer,
-		},
+	changePeer := &pdpb.ChangePeer{
+		ChangeType: changeType.Enum(),
+		Peer:       peer,
 	}
 
-	req := &raft_cmdpb.RaftCmdRequest{
-		Header: &raft_cmdpb.RaftRequestHeader{
-			RegionId:    proto.Uint64(regionID),
-			RegionEpoch: region.RegionEpoch,
-			Peer:        request.Leader,
-		},
-		AdminRequest: changePeer,
-	}
+	return changePeer, nil
+}
 
-	job := &pd_jobpd.Job{
-		Request: req,
-		Region:  request.Region,
+func (c *raftCluster) maybeChangePeer(request *pdpb.RegionHeartbeatRequest, reqRegion *metapb.Region,
+	region *metapb.Region) ([]clientv3.Op, *pdpb.ChangePeer, error) {
+	leader := request.GetLeader()
+	if leader == nil {
+		return nil, nil, errors.Errorf("invalid request leader, %v", request)
+	}
+	if region == nil {
+		return nil, nil, nil
 	}
 
-	return c.postJob(job)
-}
+	regionEpoch := region.GetRegionEpoch()
+	reqRegionEpoch := reqRegion.GetRegionEpoch()
 
-func (c *raftCluster) handleChangePeerOK(changePeer *raft_cmdpb.ChangePeerResponse) error {
-	region := changePeer.Region
+	if reqRegionEpoch.GetConfVer() < regionEpoch.GetConfVer() {
+		// If the request epoch configure version is less than the current one, return an error.
+		return nil, nil, errors.Errorf("invalid region epoch, request: %v, currenrt: %v", reqRegionEpoch, regionEpoch)
+	} else if reqRegionEpoch.GetConfVer() > regionEpoch.GetConfVer() {
+		// If the request epoch configure version is greater than the current one, update region meta.
+		regionSearchPath := makeRegionSearchKey(c.clusterRoot, reqRegion.GetEndKey())
+		regionValue, err := proto.Marshal(reqRegion)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
 
-	// Update region
-	regionSearchPath := makeRegionSearchKey(c.clusterRoot, region.GetEndKey())
-	regionValue, err := proto.Marshal(region)
-	if err != nil {
-		return errors.Trace(err)
+		var ops []clientv3.Op
+		ops = append(ops, clientv3.OpPut(regionSearchPath, string(regionValue)))
+		return ops, nil, nil
```

Comment:
E,g, pd peers (1), ConfVer (1), tikv (1, 2), ConfVer (2), when tikv reports, pd becomes (1, 2) (2), but it will let tikv do ChangPeer in next heartbeat, but I think pd can do 1) update region 2) tell tikv to ChangePeer at same time.